python/pyarrow/parquet.py

```diff
@@ -1252,6 +1341,92 @@ def _make_manifest(path_or_paths, fs, pathsep='/', metadata_nthreads=1,
     return pieces, partitions, common_metadata_path, metadata_path
 
 
+class _ParquetDatasetV2:
+    """
+    ParquetDataset shim using the Dataset API under the hood.
+    """
+    def __init__(self, path_or_paths, filesystem=None, filters=None,
+                 read_dictionary=None, buffer_size=None, memory_map=False,
+                 **kwargs):
+        import pyarrow.dataset as ds
+        import pyarrow.fs
+
+        # Raise error for not supported keywords
+        for keyword, default in [
+                ("schema", None), ("metadata", None),
+                ("split_row_groups", False), ("validate_schema", True),
+                ("metadata_nthreads", 1)]:
+            if keyword in kwargs and kwargs[keyword] is not default:
+                raise ValueError(
+                    "Keyword '{0}' is not yet supported with the new "
+                    "Dataset API".format(keyword))
+
+        # map old filesystems to new one
+        # TODO(dataset) deal with other file systems
+        if isinstance(filesystem, LocalFileSystem):
+            filesystem = pyarrow.fs.LocalFileSystem(use_mmap=memory_map)
+        elif filesystem is None and memory_map:
+            filesystem = pyarrow.fs.LocalFileSystem(use_mmap=True)
```

Comment:
```suggestion
        if filesystem is None or isinstance(filesystem, LocalFileSystem):
            filesystem = pyarrow.fs.LocalFileSystem(use_mmap=memory_map)
```

Reply:
The general problem here (which I didn't express), is that if the filesystem is not specified, it actually can be a non-local filesystem if the file path is a URI like "s3://" (that includes all connection options). So we shouldn't generally change None to a LocalFileSystem. But I was assuming that `memory_map=True` is only applicable for a LocalFileSystem (is that assumption correct?), so that if the user specified this, it should be safe to override None with a LocalFileSystem.

(if we keep the code as it is, I should certainly add a comment about that)

Reply:
I see, I had forgotten that the path might be a URI. I believe memory_map only applies to LocalFileSystem (though it may apply transitively to SubTreeFileSystem if that wraps a local fs)